File: projects/resources/python/benchmark/benchmark.py
# Copyright (c) 2020, 2021, NECSTLab, Politecnico di Milano. All rights reserved.

# Redistribution and use in source and binary forms, with or without
# modification, are permitted provided that the following conditions
# are met:
#  * Redistributions of source code must retain the above copyright
#    notice, this list of conditions and the following disclaimer.
#  * Redistributions in binary form must reproduce the above copyright
#    notice, this list of conditions and the following disclaimer in the
#    documentation and/or other materials provided with the distribution.
#  * Neither the name of NECSTLab nor the names of its
#    contributors may be used to endorse or promote products derived
#    from this software without specific prior written permission.
#  * Neither the name of Politecnico di Milano nor the names of its
#    contributors may be used to endorse or promote products derived
#    from this software without specific prior written permission.

# THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS ``AS IS'' AND ANY
# EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
# IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR
# PURPOSE ARE DISCLAIMED.  IN NO EVENT SHALL THE COPYRIGHT OWNER OR
# CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL,
# EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO,
# PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR
# PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY
# OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
# (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
# OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.

from benchmark_result import BenchmarkResult
from abc import ABC, abstractmethod
from java.lang import System
from typing import Callable
import polyglot

DEFAULT_BLOCK_SIZE_1D = 32
DEFAULT_BLOCK_SIZE_2D = 8
DEFAULT_NUM_BLOCKS = 64  # GTX 960, 8 SM
DEFAULT_NUM_BLOCKS = 448  # P100, 56 SM
DEFAULT_NUM_BLOCKS = 176  # GTX 1660 Super, 22 SM

def time_phase(phase_name: str) -> Callable:
    """
    Decorator that simplifies timing a function call and storing the result in the benchmark log;
    :param phase_name: name of the benchmark phase
    :return: the output of the wrapped function
    """
    def inner_func(func) -> Callable:
        def func_call(self, *args, **kwargs) -> object:
            start = System.nanoTime()
            result = func(self, *args, **kwargs)
            end = System.nanoTime()
            self.benchmark.add_phase({"name": phase_name, "time_sec": (end - start) / 1_000_000_000})
            return result
        return func_call
    return inner_func


class Benchmark(ABC):
    """
    Base class for all benchmarks, it provides the general control flow of the benchmark execution;
    :param name: name of the benchmark
    :param benchmark: instance of BenchmarkResult, used to store results
    :param nvprof_profile: if present activate profiling for nvprof when running the benchmark
    """

    def __init__(self, name: str, benchmark: BenchmarkResult, nvprof_profile: bool = False):
        self.name = name
        self.benchmark = benchmark
        self.nvprof_profile = nvprof_profile
        self.time_phases = False
        self.tot_iter = 0
        self.current_iter = 0
        self.random_seed = 42  # Default random seed, it will be overwritten with a random one;
        self.block_size_1d = DEFAULT_BLOCK_SIZE_1D
        self.block_size_2d = DEFAULT_BLOCK_SIZE_2D
        self.num_blocks = DEFAULT_NUM_BLOCKS
        self._block_size = {}

    @abstractmethod
    def alloc(self, size: int, block_size: dict = None) -> None:
        """
        Allocate new memory on GPU used for the benchmark;
        :param size: base factor used in the memory allocation, e.g. size of each array
        :param block_size: optional dictionary containing block size for 1D and 2D kernels
        """
        pass

    @abstractmethod
    def init(self) -> None:
        """
        Initialize the content of the input data of the benchmark;
        """
        pass

    @abstractmethod
    def reset_result(self) -> None:
        """
        Reset the values that hold the GPU result
        """
        pass

    @abstractmethod
    def cpu_validation(self, gpu_result: object, reinit: bool) -> None:
        """
        Run an equivalent benchmark on CPU to obtain the correct result of the benchmark,
        and compute the distance w.r.t. the GPU result;
        :param gpu_result: the output of the GPU computation
        :param reinit: if the GPU data was re-initialized in this computation
        """
        pass

    @abstractmethod
    def execute(self) -> object:
        """
        Execute the main computation of this benchmark;
        :return: the result of the GPU computation, it could be a scalar numeric value or an arbitrary data structure
        """
        pass

    def execute_phase(self, phase_name, function, *args) -> object:
        """
        Executes a single step of the benchmark, possibily measuring the time it takes
        :param phase_name: name of this benchmark step
        :param function: a function to execute
        :param args: arguments of the function
        :return: the result of the function
        """
        if self.time_phases:
            start = System.nanoTime()
            res = function(*args)
            end = System.nanoTime()
            self.benchmark.add_phase({"name": phase_name, "time_sec": (end - start) / 1_000_000_000})
            return res
        else:
            return function(*args)

    def run(self, num_iter: int, policy: str, size: int, realloc: bool, reinit: bool,
            time_phases: bool, block_size: dict = None, prevent_reinit=False, number_of_blocks=DEFAULT_NUM_BLOCKS) -> None:

        # Fix missing block size;
        if "block_size_1d" not in block_size:
            block_size["block_size_1d"] = DEFAULT_BLOCK_SIZE_1D
        if "block_size_2d" not in block_size:
            block_size["block_size_2d"] = DEFAULT_BLOCK_SIZE_2D
        if number_of_blocks:
            self.num_blocks = number_of_blocks

        self.benchmark.start_new_benchmark(name=self.name,
                                           policy=policy,
                                           size=size,
                                           realloc=realloc,
                                           reinit=reinit,
                                           block_size=block_size,
                                           iteration=num_iter,
                                           time_phases=time_phases)
        self.current_iter = num_iter
        self.time_phases = time_phases
        self._block_size = block_size
        # TODO: set the execution policy;

        # Start a timer to monitor the total GPU execution time;
        start = System.nanoTime()

        # Allocate memory for the benchmark;
        if (num_iter == 0 or realloc) and not prevent_reinit:
            self.alloc(size, block_size)
        # Initialize memory for the benchmark;
        if (num_iter == 0 or reinit) and not prevent_reinit:
            self.init()

        # Reset the result;
        self.reset_result()

        # Start nvprof profiling if required;
        if self.nvprof_profile:
            polyglot.eval(language="grcuda", string="cudaProfilerStart")()

        # Execute the benchmark;
        gpu_result = self.execute()

        # Stop nvprof profiling if required;
        if self.nvprof_profile:
            polyglot.eval(language="grcuda", string="cudaProfilerStop")()

        # Stop the timer;
        end = System.nanoTime()
        self.benchmark.add_total_time((end - start) / 1_000_000_000)

        # Perform validation on CPU;
        if self.benchmark.cpu_validation:
            self.cpu_validation(gpu_result, reinit)

        # Write to file the current result;
        self.benchmark.save_to_file()
        # Book-keeping;
        self.tot_iter += 1

File: projects/resources/java/grcuda-benchmark/src/main/java/it/necst/grcuda/benchmark/Benchmark.java
/*
 * Copyright (c) 2022 NECSTLab, Politecnico di Milano. All rights reserved.
 *
 * Redistribution and use in source and binary forms, with or without
 * modification, are permitted provided that the following conditions
 * are met:
 *  * Redistributions of source code must retain the above copyright
 *    notice, this list of conditions and the following disclaimer.
 *  * Redistributions in binary form must reproduce the above copyright
 *    notice, this list of conditions and the following disclaimer in the
 *    documentation and/or other materials provided with the distribution.
 *  * Neither the name of NECSTLab nor the names of its
 *    contributors may be used to endorse or promote products derived
 *    from this software without specific prior written permission.
 *  * Neither the name of Politecnico di Milano nor the names of its
 *    contributors may be used to endorse or promote products derived
 *    from this software without specific prior written permission.
 *
 * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS ``AS IS'' AND ANY
 * EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
 * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR
 * PURPOSE ARE DISCLAIMED.  IN NO EVENT SHALL THE COPYRIGHT OWNER OR
 * CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL,
 * EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO,
 * PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR
 * PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY
 * OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
 * (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
 * OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
 */

package it.necst.grcuda.benchmark;

import java.util.ArrayList;
import java.util.function.Consumer;
import org.graalvm.polyglot.Context;
import org.graalvm.polyglot.Value;


public abstract class Benchmark {
    public Context context;
    public final BenchmarkConfig config;
    public final BenchmarkResults benchmarkResults;
    public ArrayList<Value> deviceArrayList = new ArrayList<>(); // used to store all the arrays to be freed at the end of the benchmark

    public Benchmark(BenchmarkConfig currentConfig) {
        this.config = currentConfig;
        this.benchmarkResults = new BenchmarkResults(currentConfig);
        this.context = createContext(currentConfig);
    }

    /**
     * This method is used to run the current benchmark.
     * It will use the information stored in the config attribute to decide whether to do an additional initialization phase and
     the cpuValidation.
     */
    public void run() {
        if(config.debug)
            System.out.println("INSIDE run()");

        for (int i = 0; i < config.totIter; ++i) {
            if(config.debug)
                System.out.println("["+i+"] START");
            benchmarkResults.startNewIteration(i, config.timePhases); // create the current iteration in the result class

            // Start a timer to monitor the total GPU execution time
            long overall_gpu_start = System.nanoTime();

            // Allocate memory for the benchmark

            if (config.reAlloc || i == 0){
                if(config.debug)
                    System.out.println("["+i+"] alloc");
                time(i, "alloc", this::allocateTest);
            }

            // Initialize memory for the benchmark

            if (config.reInit || i == 0){
                if(config.debug)
                    System.out.println("["+i+"] init");
                time(i, "init", this::initializeTest);
            }

            // Reset the result
            if(config.debug)
                System.out.println("["+i+"] reset");
            time(i, "reset", this::resetIteration);

            if(config.nvprof_profile){
                context.eval("grcuda", "cudaProfilerStart").execute();
            }

            // Execute the benchmark
            if(config.debug)
                System.out.println("["+i+"] execution");
            time(i, "execution", this::runTest);

            if(config.nvprof_profile){
                context.eval("grcuda", "cudaProfilerStop").execute();
            }

            // Stop the timer
            long overall_gpu_end = System.nanoTime();

            benchmarkResults.setCurrentTotalTime((overall_gpu_end - overall_gpu_start) / 1000000000F);

            // Perform validation on CPU
            if (config.cpuValidate && i == 0)
                cpuValidation();

            if(config.debug)
                System.out.println("["+i+"] VALIDATION \nCPU: " + benchmarkResults.cpu_result+"\nGPU: " + benchmarkResults.currentIteration().gpu_result);
        }

        // Save the benchmark results
        benchmarkResults.saveToJsonFile();


        // Free the allocated arrays
        deallocDeviceArrays();

        //  Gracefully close the current context
        context.close();
    }

    /**
     * This method is used to time the function passed to it.
     * It will add the timing and the phase name to the benchmarkResult attribute.
     * @param iteration the current iteration of the benchmark
     * @param phaseName the current phase of the benchmark
     * @param functionToTime the function to time passed like "class::funName"
     */
    private void time(int iteration, String phaseName, Consumer<Integer> functionToTime){
        long begin = System.nanoTime();
        functionToTime.accept(iteration);
        long end = System.nanoTime();
        benchmarkResults.addPhaseToCurrentIteration(phaseName, (end - begin)/ 1000000000F);
    }

    protected void deallocDeviceArrays(){
        for(Value v : deviceArrayList)
            v.invokeMember("free");
    }

    protected Value requestArray(String type, int size){
        Value vector = context.eval("grcuda", type+"["+ size +"]");
        deviceArrayList.add(vector);
        return vector;
    }

    private Context createContext(BenchmarkConfig config){
        return Context
                .newBuilder()
                .allowAllAccess(true)
                .allowExperimentalOptions(true)
                //logging settings
                .option("log.grcuda.com.nvidia.grcuda.level", "WARNING")
                .option("log.grcuda.com.nvidia.grcuda.GrCUDAContext.level", "SEVERE")
                //GrCUDA env settings
                .option("grcuda.ExecutionPolicy", config.executionPolicy)
                .option("grcuda.InputPrefetch", String.valueOf(config.inputPrefetch))
                .option("grcuda.RetrieveNewStreamPolicy", config.retrieveNewStreamPolicy)
                .option("grcuda.RetrieveParentStreamPolicy", config.retrieveParentStreamPolicy)
                .option("grcuda.DependencyPolicy", config.dependencyPolicy)
                .option("grcuda.DeviceSelectionPolicy", config.deviceSelectionPolicy)
                .option("grcuda.ForceStreamAttach", String.valueOf(config.forceStreamAttach))
                .option("grcuda.EnableComputationTimers", String.valueOf(config.enableComputationTimers))
                .option("grcuda.MemAdvisePolicy", config.memAdvisePolicy)
                .option("grcuda.NumberOfGPUs", String.valueOf(config.numGpus))
                .option("grcuda.BandwidthMatrix", config.bandwidthMatrix)
                .build();
    }

    /*
        ###################################################################################
                        METHODS TO BE IMPLEMENTED IN THE BENCHMARKS
        ###################################################################################
    */

    /**
     * Here goes the read of the test parameters,
     * the initialization of the necessary arrays
     * and the creation of the kernels (if applicable)
     * @param iteration the current number of the iteration
     */
    protected abstract void initializeTest(int iteration);

    /**
     * Allocate new memory on GPU used for the benchmark
     * @param iteration the current number of the iteration
     */
    protected abstract void allocateTest(int iteration);

    /**
     * Reset code, to be run before each test
     * Here you clean up the arrays and other reset stuffs
     * @param iteration the current number of the iteration
     */
    protected abstract void resetIteration(int iteration);

    /**
     * Run the actual test
     * @param iteration the current number of the iteration
     */
    protected abstract void runTest(int iteration);

    /**
     * (numerically) validate results against CPU
     */
    protected abstract void cpuValidation();

}
